### PR TITLE
Restore action for cluster/namespace role bindings

### PR DESCRIPTION
**Pull request by alexander-demichev**
Fixes https://github.com/vmware-tanzu/velero/issues/80

### CODE REVIEW

File: pkg/restore/clusterrolebinding_action.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	if len(namespaceMapping) == 0 {
+		return velero.NewRestoreItemActionExecuteOutput(&unstructured.Unstructured{Object: input.Item.UnstructuredContent()}), nil
+	}
+
+	clusterRoleBinding := new(rbac.ClusterRoleBinding)
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(input.Item.UnstructuredContent(), clusterRoleBinding); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	a.logger.Infoln(123456)
```
Comment (dymurray) [inline pkg/restore/clusterrolebinding_action.go]: What does this logger statement do? Seems odd.

File: pkg/restore/clusterrolebinding_action_test.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	require.NoError(t, err)
+	assert.Equal(t, velero.ResourceSelector{IncludedResources: []string{"clusterrolebindings"}}, actual)
+}
+
+// 3 cases: no mapping - return rolebinding unchanged, mapping enabled change namespace
+func TestClusterRoleBindingActionExecute(t *testing.T) {
+	tests := []struct {
+		name             string
+		namespaces       []string
+		namespacemapping map[string]string
```
Comment (prydonius) [inline pkg/restore/clusterrolebinding_action_test.go]: naming nit:
```suggestion
		namespaceMapping map[string]string
```

File: pkg/restore/clusterrolebinding_action_test.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+)
+
+func TestClusterRoleBindingActionAppliesTo(t *testing.T) {
+	action := NewClusterRoleBindingAction(test.NewLogger())
+	actual, err := action.AppliesTo()
+	require.NoError(t, err)
+	assert.Equal(t, velero.ResourceSelector{IncludedResources: []string{"clusterrolebindings"}}, actual)
+}
+
+// 3 cases: no mapping - return rolebinding unchanged, mapping enabled change namespace
```
Comment (prydonius) [inline pkg/restore/clusterrolebinding_action_test.go]: 2 cases? I don't think we really need this comment :)

File: pkg/restore/clusterrolebinding_action_test.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	}{
+		{
+			name:             "namespace mapping disabled",
+			namespaces:       []string{"foo"},
+			namespacemapping: map[string]string{},
+			expected:         []string{"foo"},
+		},
+		{
+			name:             "namespace mapping enabled",
+			namespaces:       []string{"foo"},
```
Comment (prydonius) [inline pkg/restore/clusterrolebinding_action_test.go]: It would be good to test that a namespace not included in the mapping is left unchanged.

File: pkg/restore/rolebinding_action.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+func (a *RoleBindingAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"rolebindings"},
+	}, nil
+}
+
+func (a *RoleBindingAction) Execute(input *velero.RestoreItemActionExecuteInput) (*velero.RestoreItemActionExecuteOutput, error) {
+	namespaceMapping := input.Restore.Spec.NamespaceMapping
```
[the remark this answers is not in the document]
Comment (alexander-demichev) [reply pkg/restore/rolebinding_action.go]: If you have any ideas, I'll be happy to hear them.

**Comment by alexander-demichev** [comment]
@prydonius @dymurray @skriss Thank you all for reviews! Sorry for debugging prints/comments, I should've had more coffee and sleep yesterday:)

File: pkg/restore/rolebinding_action.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (skriss) [reply pkg/restore/rolebinding_action.go]: seems like L54-59 are really the only common ones - you could extract them to a helper func, like:

```go
func remapSubjectNamespaces(subjects []rbac.Subject, namespaceMappings map[string]string) []rbac.Subject {
    // ...
}
```

I'm fine either way, as it's only a couple lines of simple code.

File: pkg/restore/rolebinding_action_test.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2018, 2019 the Velero contributors.
```
Comment (skriss) [inline pkg/restore/rolebinding_action_test.go]: 2019

File: pkg/restore/rolebinding_action_test.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+func TestRoleBindingActionAppliesTo(t *testing.T) {
+	action := NewRoleBindingAction(test.NewLogger())
+	actual, err := action.AppliesTo()
+	require.NoError(t, err)
+	assert.Equal(t, velero.ResourceSelector{IncludedResources: []string{"rolebindings"}}, actual)
+}
+
+func TestRoleBindingActionExecute(t *testing.T) {
+	tests := []struct {
```
Comment (skriss) [inline pkg/restore/rolebinding_action_test.go]: could we add a test case where namespace mappings are non-empty, but none of them apply to the role binding?

**Comment by skriss** [comment]
One other thing - please add a changelog (`changelogs/unreleased/1974-alexander-demichev`)

File: pkg/restore/clusterrolebinding_action_test.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+	"github.com/vmware-tanzu/velero/pkg/test"
+)
+
+func TestClusterRoleBindingActionAppliesTo(t *testing.T) {
+	action := NewClusterRoleBindingAction(test.NewLogger())
+	actual, err := action.AppliesTo()
+	require.NoError(t, err)
+	assert.Equal(t, velero.ResourceSelector{IncludedResources: []string{"clusterrolebindings"}}, actual)
+}
```
Comment (prydonius) [inline pkg/restore/clusterrolebinding_action_test.go]: new line?

Comment (alexander-demichev) [reply pkg/restore/clusterrolebinding_action_test.go]: Thanks, fixed

**Comment by skriss** [comment]
@alexander-demichev guessing you may need to rebase on top of the latest `master` code

**Comment by prydonius** [comment]
@skriss I just ran this locally and it looks like it's because controller-gen is now adding an annotation:
```
+  annotations:
+    controller-gen.kubebuilder.io/version: (unknown)
```
to all CRDs. Should we fix the version we're pulling in in the builder image so we're not randomly affected by future controller-gen changes?

**Comment by skriss** [comment]
Ah. yeah let's pin the version we're using

**Comment by prydonius** [comment]
Ah, we needed to use master to get the single-version CRD fix, but this is now in the latest release, so we can now fix to that. I'll send a PR to fix this.

**Comment by prydonius** [comment]
Restarted the test and this is now passing 🎉

**Comment by skriss** [comment]
Looks like we're good to merge then
